src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamValueTaskSource.cs

```diff
@@ -0,0 +1,252 @@
+            }
+
+            private void CompleteCallback(ulong packedResult)
+            {
+                // Free up the native resource and cancellation registration
+                CancellationToken cancellationToken = _cancellationRegistration.Token; // access before disposing registration
+                ReleaseNativeResource();
+
+                // Unpack the result and send it to the user
+                long result = (long)(packedResult & FileStreamHelpers.ResultMask);
+                if (result == FileStreamHelpers.ResultError)
+                {
+                    int errorCode = unchecked((int)(packedResult & uint.MaxValue));
+                    if (errorCode == Interop.Errors.ERROR_OPERATION_ABORTED)
+                    {
+                        _source.SetException(ExceptionDispatchInfo.SetCurrentStackTrace(new OperationCanceledException(cancellationToken.IsCancellationRequested ? cancellationToken : new CancellationToken(canceled: true))));
+                    }
+                    else
+                    {
+                        Exception e = Win32Marshal.GetExceptionForWin32Error(errorCode);
+                        e.SetCurrentStackTrace();
+                        _source.SetException(ExceptionDispatchInfo.SetCurrentStackTrace(e));
+                    }
+                }
+                else
+                {
+                    Debug.Assert(result == FileStreamHelpers.ResultSuccess, "Unknown result");
+                    _source.SetResult((int)(packedResult & uint.MaxValue) + _numBufferedBytes);
+                }
+            }
+
+            private void Cancel(CancellationToken token)
+            {
+                // If the handle is still valid, attempt to cancel the IO
+                if (!_strategy._fileHandle.IsInvalid &&
+                    !Interop.Kernel32.CancelIoEx(_strategy._fileHandle, _overlapped))
+                {
+                    int errorCode = Marshal.GetLastWin32Error();
+
+                    // ERROR_NOT_FOUND is returned if CancelIoEx cannot find the request to cancel.
+                    // This probably means that the IO operation has completed.
+                    if (errorCode != Interop.Errors.ERROR_NOT_FOUND)
+                    {
+                        _source.SetException(ExceptionDispatchInfo.SetCurrentStackTrace( // TODO: Resource string for exception
+                            new OperationCanceledException("IO operation cancelled.", Win32Marshal.GetExceptionForWin32Error(errorCode), token)));
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Extends <see cref="FileStreamValueTaskSource"/> with to support disposing of a
+        /// <see cref="MemoryHandle"/> when the operation has completed.  This should only be used
+        /// when memory doesn't wrap a byte[].
+        /// </summary>
+        private sealed class MemoryAwaitableProvider : FileStreamValueTaskSource
```

Comment:
> This should only be used when memory doesn't wrap a byte[].

I just realized that we don't have benchmarks for this scenario as `Memory` used in our micro-benchmarks always wraps a byte array. Could you please add such benchmarks to https://github.com/dotnet/performance/blob/main/src/benchmarks/micro/libraries/System.IO.FileSystem/Perf.FileStream.cs ? One test case should be enough as this should be rare.

Reply:
Some of the benchmarks do consume this type. If you look at some of the profiling tests I executed, you'll see that `MemoryFileStreamValueTaskSource` shows up a few times. The profiling tests reuse the code we had in our benchmarks.

I don't mind adding an explicit one though.